File: packages/path_core_dart/test/adoption_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:path_core_dart/path_core_dart.dart';
import 'package:test/test.dart';

import 'path_repository_test.dart' show MemoryStorage;

void main() {
  test('queued command snapshots match their durable intent', () async {
    final repo = PathRepository(MemoryStorage());
    await repo.load();
    final tags = ['original'];
    final creation = repo.createEntity(
      commandId: 'snapshot',
      id: 'snapshot',
      type: 'dream',
      title: 'dream',
      tags: tags,
    );
    tags.add('too-late');
    await creation;
    expect(repo.document.entities.single['tags'], ['original']);
    final patch = <String, dynamic>{'why': 'original'};
    final update = repo.updateEntity(
      commandId: 'snapshot-update',
      entityId: 'snapshot',
      patch: patch,
    );
    patch['why'] = 'too-late';
    await update;
    expect(repo.document.entities.single['why'], 'original');
    await repo.updateEntity(
      commandId: 'snapshot-update',
      entityId: 'snapshot',
      patch: {'why': 'original'},
    );
    expect(repo.document.revision, 2);
  });
  final fixture = jsonDecode(
    File('../path-core/fixtures/adoption-conformance-v1.json')
        .readAsStringSync(),
  ) as Map<String, dynamic>;
  for (final item in fixture['cases'] as List) {
    test('shared conformance: ${item['name']}', () {
      if (item['valid'] == true) {
        final document = PathDocument.fromJson(
          item['document'] as Map<String, dynamic>,
        );
        expect(jsonDecode(document.encode()), item['document']);
      } else {
        expect(
          () => PathDocument.fromJson(item['document'] as Map<String, dynamic>),
          throwsFormatException,
        );
      }
    });
  }
  test('all creation types, nested actions, why, parent restrictions and historical parents', () async {
    final repo = PathRepository(MemoryStorage());
    await repo.load();
    for (final type in pathEntityTypes.where((t) => t != 'focus-session')) {
      await repo.createEntity(
        commandId: 'create-$type',
        id: type,
        type: type,
        title: type,
        why: 'meaning',
      );
    }
    await repo.createEntity(
      commandId: 'nested',
      id: 'nested',
      type: 'action',
      title: 'nested',
      parentId: 'action',
    );
    expect(repo.document.entities.last['parentId'], 'action');
    await expectLater(
      repo.reparent(commandId: 'cycle', entityId: 'action', parentId: 'nested'),
      throwsA(isA<PathCommandException>()),
    );
    await expectLater(
      repo.createEntity(
        commandId: 'invalid',
        type: 'dream',
        title: 'invalid',
        parentId: 'goal',
      ),
      throwsA(isA<PathCommandException>()),
    );
    final historical = repo.document;
    historical.entities.first['parentId'] = 'missing';
    expect(
      PathDocument.fromJson(historical.json).entities.first['parentId'],
      'missing',
    );
    await expectLater(
      repo.createEntity(
        commandId: 'missing',
        type: 'goal',
        title: 'invalid',
        parentId: 'missing',
      ),
      throwsA(isA<PathCommandException>()),
    );
  });
  test(
    'every command replays durably and rejects changed intent with no write',
    () async {
      final storage = MemoryStorage();
      var repo = PathRepository(storage);
      await repo.load();
      await repo.createEntity(
        commandId: 'goal',
        id: 'goal',
        type: 'goal',
        title: 'goal',
      );
      await repo.createEntity(
        commandId: 'action',
        id: 'action',
        type: 'action',
        title: 'action',
        extensions: {'future': true},
      );
      Future<void> replay(
        Future<Map<String, dynamic>> Function(PathRepository) command,
      ) async {
        await command(repo);
        final persisted = storage.contents;
        repo = PathRepository(storage);
        await repo.load();
        await command(repo);
        expect(storage.contents, persisted);
      }

      await replay(
        (r) => r.updateEntity(
          commandId: 'update',
          entityId: 'action',
          patch: {
            'why': 'important',
            'extensions': {'new': 1},
          },
        ),
      );
      expect(repo.document.entities.last['extensions'], {
        'future': true,
        'new': 1,
      });
      await replay(
        (r) =>
            r.plan(commandId: 'plan', entityId: 'action', date: '2026-09-08'),
      );
      await replay((r) => r.complete(commandId: 'done', entityId: 'action'));
      await replay((r) => r.reopen(commandId: 'reopen', entityId: 'action'));
      await replay(
        (r) => r.reparent(
          commandId: 'parent',
          entityId: 'action',
          parentId: 'goal',
        ),
      );
      await replay(
        (r) => r.startFocusSession(
          commandId: 'start',
          actionId: 'action',
          id: 'focus',
          mode: 'focus',
        ),
      );
      await replay(
        (r) => r.endFocusSession(
          commandId: 'end',
          entityId: 'focus',
          outcome: 'completed',
          handoffNote: 'next',
        ),
      );
      await replay(
        (r) => r.deleteEntity(commandId: 'delete-focus', entityId: 'focus'),
      );
      await replay(
        (r) => r.deleteEntity(commandId: 'delete', entityId: 'action'),
      );
      final persisted = storage.contents;
      await expectLater(
        repo.plan(commandId: 'plan', entityId: 'action', date: '2026-09-09'),
        throwsA(isA<PathCommandException>()),
      );
      expect(storage.contents, persisted);
      await expectLater(
        repo.createEntity(
          commandId: 'dead-parent',
          type: 'action',
          title: 'bad',
          parentId: 'action',
        ),
        throwsA(isA<PathCommandException>()),
      );
    },
  );
  test('failed writes recover, invalid dates and readonly snapshots cannot corrupt the repository', () async {
    final storage = MemoryStorage();
    final repo = PathRepository(storage);
    await repo.load();
    await repo.createEntity(
      commandId: 'a',
      id: 'a',
      type: 'action',
      title: 'a',
    );
    final original = storage.contents;
    await expectLater(
      repo.plan(commandId: 'invalid-date', entityId: 'a', date: '2026-02-30'),
      throwsA(isA<PathCommandException>()),
    );
    repo.document.entities.clear();
    expect(repo.document.entities.length, 1);
    storage.failWrites = true;
    await expectLater(
      repo.updateEntity(
        commandId: 'retry',
        entityId: 'a',
        patch: {'title': 'b'},
      ),
      throwsStateError,
    );
    expect(storage.contents, original);
    expect(repo.document.entities.single['title'], 'a');
    storage.failWrites = false;
    await repo.updateEntity(
      commandId: 'retry',
      entityId: 'a',
      patch: {'title': 'b'},
    );
    expect(repo.document.entities.single['title'], 'b');
  });
  test('cannot delete live parent or create ID collision', () async {
    final repo = PathRepository(MemoryStorage());
    await repo.load();
    await repo.createEntity(
      commandId: 'a',
      id: 'a',
      type: 'action',
      title: 'a',
    );
    await repo.createEntity(
      commandId: 'b',
      id: 'b',
      type: 'action',
      title: 'b',
      parentId: 'a',
    );
    await expectLater(
      repo.deleteEntity(commandId: 'delete', entityId: 'a'),
      throwsA(isA<PathCommandException>()),
    );
    await expectLater(
      repo.createEntity(
        commandId: 'collision',
        id: 'a',
        type: 'goal',
        title: 'a',
      ),
      throwsA(isA<PathCommandException>()),
    );
    expect(repo.document.revision, 2);
  });
}

File: packages/path-core/tests/adoption.test.mjs
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import test from 'node:test';
import { decodeCanonical, PathRepository } from '../src/repository.ts';
import { createPathCommandPort } from '../src/command-port.ts';

const fixture = JSON.parse(await readFile(new URL('../fixtures/adoption-conformance-v1.json', import.meta.url), 'utf8'));
for (const entry of fixture.cases) test(`conformance: ${entry.name}`, () => {
    if (entry.valid) assert.deepEqual(decodeCanonical(entry.document), entry.document);
    else assert.throws(() => { if (!decodeCanonical(entry.document)) throw new Error('unsupported'); });
});

function harness() {
    let stored = { repositoryVersion: 1, envelope: { schemaVersion: 1, revision: 0, entities: [], events: [], extensions: { future: true } }, settings: { unknown: 'retain' }, extensions: { future: [1, 2] } };
    let writes = 0, sequence = 0;
    const repository = new PathRepository({ load: async () => stored, save: async value => { stored = value; writes++; } });
    const ready = repository.load();
    const port = createPathCommandPort({
        async updateDocument(update) { await ready; const next = await update(structuredClone(stored)); return next ? repository.save(next, stored.envelope.revision) : structuredClone(stored); },
        afterPersist() {}, now: () => '2026-09-06T12:00:00Z', createId: () => `event-${++sequence}`,
    });
    return { port, get document() { return stored; }, get writes() { return writes; } };
}

test('every product entity can be created, edited, and retained with its meaning and future fields', async () => {
    const h = harness();
    const types = ['dream', 'goal', 'milestone', 'action', 'habit', 'evidence', 'reflection'];
    const parents = [undefined, 'dream', 'goal', 'milestone', 'goal', 'action', 'dream'];
    for (const [index, type] of types.entries()) {
        const command = { type: 'create-entity', commandId: `create-${type}`, input: { id: type, type, title: type, ...(parents[index] ? { parentId: parents[index] } : {}), extensions: { future: { keep: 42 } }, ...(type === 'dream' ? { why: 'Meaning' } : {}) } };
        assert.equal((await h.port.execute(command)).accepted, true);
        assert.equal((await h.port.execute(command)).replayed, true);
        assert.equal((await h.port.execute({ ...command, input: { ...command.input, title: 'different' } })).accepted, false);
        assert.equal((await h.port.execute({ type: 'update-entity', commandId: `update-${type}`, entityId: type, patch: { description: 'Changed', extensions: { known: true } } })).accepted, true);
    }
    assert.equal(h.writes, 14);
    assert.equal(h.document.envelope.revision, 14);
    assert.equal(h.document.envelope.entities[0].why, 'Meaning');
    for (const entity of h.document.envelope.entities) assert.deepEqual(entity.extensions.future, { keep: 42 });
    assert.deepEqual(h.document.extensions, { future: [1, 2] });
    assert.deepEqual(decodeCanonical(JSON.parse(JSON.stringify(h.document))), h.document);
});

test('nested actions preserve hierarchy, reject cycles and cannot mutate a tombstone', async () => {
    const h = harness();
    for (const [id, parentId] of [['parent', undefined], ['child', 'parent']]) {
        assert.equal((await h.port.execute({ type: 'create-entity', commandId: id, input: { id, type: 'action', title: id, ...(parentId ? { parentId } : {}) } })).accepted, true);
    }
    assert.deepEqual(await h.port.execute({ type: 'reparent', commandId: 'cycle', entityId: 'parent', nextParentId: 'child' }), { accepted: false, reason: 'cycle' });
    assert.equal((await h.port.execute({ type: 'delete-entity', commandId: 'delete-parent', entityId: 'parent' })).reason, 'has-children');
    const remove = { type: 'delete-entity', commandId: 'delete-child', entityId: 'child' };
    assert.equal((await h.port.execute(remove)).accepted, true);
    assert.equal((await h.port.execute(remove)).replayed, true);
    for (const type of ['complete', 'reopen']) assert.equal((await h.port.execute({ type, commandId: type, entityId: 'child' })).accepted, false);
    assert.equal((await h.port.execute({ type: 'create-entity', commandId: 'under-deleted', input: { id: 'orphan', type: 'action', title: 'Orphan', parentId: 'child' } })).accepted, false);
    assert.equal(h.document.envelope.entities.length, 2);
    assert.equal(h.document.envelope.entities[1].parentId, 'parent');
    assert.equal(h.writes, 3);
});

test('planning, completion, reopening and reparenting replay without extra writes', async () => {
    const h = harness();
    await h.port.execute({ type: 'create-entity', commandId: 'create', input: { id: 'a', type: 'action', title: 'A' } });
    const commands = [
        { type: 'schedule', planned: { start: '2026-09-06' } },
        { type: 'reschedule', planned: { start: '2026-09-07' } },
        { type: 'complete' }, { type: 'reopen' },
    ];
    for (const command of commands) {
        const input = { ...command, entityId: 'a', commandId: command.type };
        assert.equal((await h.port.execute(input)).accepted, true);
        const revision = h.document.envelope.revision;
        assert.equal((await h.port.execute(input)).replayed, true);
        assert.equal(h.document.envelope.revision, revision);
    }
    assert.equal((await h.port.execute({ type: 'reschedule', commandId: 'bad-date', entityId: 'a', planned: { start: '2026-02-30' } })).reason, 'invalid-date');
    assert.equal(h.writes, 5);
});

test('legacy inverted/mixed periods remain recoverable and can be checkpointed', async () => {
    for (const end of ['2026-09-01', '2026-09-03T12:00:00Z']) {
        let stored = { goals: [{ id: 'legacy', title: 'Legacy', startDate: '2026-09-02', dueDate: end }] };
        const repository = new PathRepository({ load: async () => stored, save: async value => { stored = value; } });
        const loaded = await repository.load();
        assert.equal(loaded.migrated, true);
        assert.equal(loaded.document.envelope.entities[0].planned, undefined);
        assert.equal(loaded.document.envelope.entities[0].extensions.legacy.invalidPlanned.end, end);
        assert.equal(loaded.document.envelope.entities[0].extensions.legacy.fields.dueDate, end);
        assert.ok(loaded.diagnostics.some(item => item.code === 'invalid-period-preserved'));
        await repository.save(loaded.document, loaded.document.envelope.revision);
        assert.deepEqual((await repository.load()).document, stored);
    }
});

test('old v1 migration periods recover once with originals, other corruption fails closed', async () => {
    let stored = structuredClone(fixture.cases[0].document);
    const entity = stored.envelope.entities[0];
    entity.planned = { start: '2026-09-07', end: '2026-09-06', future: 42 };
    entity.extensions.legacy = { kind: 'goal', fields: { startDate: '2026-09-07', dueDate: '2026-09-06' } };
    const original = structuredClone(entity.planned);
    const repository = new PathRepository({ load: async () => stored, save: async value => { stored = value; } });
    const loaded = await repository.load();
    assert.equal(loaded.migrated, true);
    assert.deepEqual(loaded.document.envelope.entities[0].extensions.legacy.invalidPlanned, original);
    await repository.save(loaded.document, loaded.document.envelope.revision);
    assert.equal((await repository.load()).migrated, false);
    stored.envelope.entities[0].planned = original;
    delete stored.envelope.entities[0].extensions.legacy;
    await assert.rejects(repository.load());
});

test('switching planned date kinds retains extensions, removes obsolete bounds and journals exact new period', async () => {
    const h = harness();
    await h.port.execute({ type: 'create-entity', commandId: 'create', input: { id: 'a', type: 'action', title: 'A', planned: { start: '2026-09-01', end: '2026-09-02', future: { keep: true } } } });
    assert.equal((await h.port.execute({ type: 'reschedule', commandId: 'move', entityId: 'a', planned: { start: '2026-09-06T12:00:00Z' } })).accepted, true);
    assert.deepEqual(h.document.envelope.entities[0].planned, { start: '2026-09-06T12:00:00Z', future: { keep: true } });
    assert.deepEqual(h.document.envelope.events.at(-1).nextPlanned, h.document.envelope.entities[0].planned);
});

test('reopening an old Focus session cannot create two active sessions', async () => {
    const h = harness();
    const commands = [
        { type: 'create-entity', input: { id: 'a', type: 'action', title: '  Keep title spacing  ' } },
        { type: 'start-focus-session', input: { id: 'f1', actionId: 'a', mode: 'focus' } },
        { type: 'end-focus-session', entityId: 'f1', input: { outcome: 'completed' } },
        { type: 'start-focus-session', input: { id: 'f2', actionId: 'a', mode: 'focus' } },
    ];
    for (const [index, command] of commands.entries()) assert.equal((await h.port.execute({ ...command, commandId: `focus-${index}` })).accepted, true);
    assert.equal((await h.port.execute({ type: 'reopen', entityId: 'f1', commandId: 'reopen' })).reason, 'active-session-exists');
    assert.equal(h.document.envelope.entities[0].title, '  Keep title spacing  ');
    assert.equal(h.document.envelope.entities.filter(e => e.type === 'focus-session' && e.status === 'in-progress').length, 1);
});
